### ISSUE TITLE
Is the ground truth contained in the input data?

### ISSUE THREAD

**Issue by LiaoLW** [mlpotter/Transformer_Time_Series#2]
Hello, mlpotter, your implementation of Enhancing the Locality and Breaking the Memory Bottleneck of Transformer on Time Series Forecasting is so much helpful, but I have some questions about the training and test part.
It seems that the ground truth is contained in the input data when training the model, is it reasonable? When testing and validating the model, will it be considered as cheating?
thanks for your code again.

**Comment by mlpotter** [comment]
Hello,

Yes the ground truth is contained in the input data. I used the training method called "Teacher Forcing", where the actual output from the training dataset (ground truth) at the current time step y(t) is used as the input in the next time step x(t+1), rather than the output generated by the network. This is a pretty common technique for training LSTMs, or any time series models.

To remain consistent for the training, validation, and testing, I used this approach to evaluate all the losses. The network is not seeing the ground truth y(t+1) and predicting y(t+1) at the same time though. If you were not to use "Teacher Forcing" for the validation and testing however, I do not think the performance would drop dramatically. I just was trying to get the proof-of-concept and architecture down for the paper. It is not bad or "incorrect" approach, but best approach would be to not use "Teaching Forcing" for the validation/test.

**Comment by LiaoLW** [comment]
> Hello,
> 
> Yes the ground truth is contained in the input data. I used the training method called "Teacher Forcing", where the actual output from the training dataset (ground truth) at the current time step y(t) is used as the input in the next time step x(t+1), rather than the output generated by the network. This is a pretty common technique for training LSTMs, or any time series models.
> 
> To remain consistent for the training, validation, and testing, I used this approach to evaluate all the losses. The network is not seeing the ground truth y(t+1) and predicting y(t+1) at the same time though. If you were not to use "Teacher Forcing" for the validation and testing however, I do not think the performance would drop dramatically. I just was trying to get the proof-of-concept and architecture down for the paper. It is not bad or "incorrect" approach, but best approach would be to not use "Teaching Forcing" for the validation/test.

 thanks for your reply, I'll try to learn more about teacher forcing.

**Comment by LiaoLW** [comment]
> Hello,
> 
> Yes the ground truth is contained in the input data. I used the training method called "Teacher Forcing", where the actual output from the training dataset (ground truth) at the current time step y(t) is used as the input in the next time step x(t+1), rather than the output generated by the network. This is a pretty common technique for training LSTMs, or any time series models.
> 
> To remain consistent for the training, validation, and testing, I used this approach to evaluate all the losses. The network is not seeing the ground truth y(t+1) and predicting y(t+1) at the same time though. If you were not to use "Teacher Forcing" for the validation and testing however, I do not think the performance would drop dramatically. I just was trying to get the proof-of-concept and architecture down for the paper. It is not bad or "incorrect" approach, but best approach would be to not use "Teaching Forcing" for the validation/test.

Sorry to bother but I'm still confusing about teacher forcing. Where is the corresponding part of teacher forcing in your codes? I can only see inputing data with ground truth into the model with a mask blocking out the future data, and when I'm trying to do this, this will lead to serious problem that mean average error becomes extremely small due to the model have seen the ground truth. Could you please give me some more detail about how you implemented teacher forcing or change it into autoregressive mode?
So much thanks.
